<!-- pr-review-bot -->

## Code Review：feat(acp): use bundled bun for npx-style connections (#2421)

### 变更概述

本 PR 将所有 ACP/MCP 的 npx 风格启动（ACP bridge、legacy `npx ...` agent 命令、auth 刷新流程、MCP stdio 传输）统一切换到 bundled bun 运行时（`bun x --bun`），并删除了不再需要的 npm 缓存恢复分支（stale cache、corrupted cache、ENOTEMPTY retry 等）。主要改动在 `acpConnectors.ts`、`AcpConnection.ts`、`McpProtocol.ts` 和 `shellEnv.ts`。

---

### 方案评估

**结论**：✅ 方案合理

通过切换到 bundled bun，所有 npx 相关的缓存腐败/版本不匹配问题从根本上消失，无需再维护 cache clean/retry 逻辑。`normalizeNpxArgsForBundledBun` 负责过滤 npx-specific 标志（`--yes`、`--prefer-offline`），`resolveNpxPath` 简化为直接返回 bundled bun 路径。方案与项目架构一致，简化效果明显。测试覆盖了关键路径（Windows UTF-8、detached process、npx arg 转换等）。

---

### 问题清单

#### 🟡 MEDIUM — 测试中 directInvoke 传递给已不再接受该属性的函数，TypeScript 报 excess property 错误

**文件**：`tests/unit/acpConnectors.test.ts`，第 145–155 行、第 164–174 行

**问题代码**：

```ts
spawnNpxBackend('claude', '@pkg/cli@1.0.0', 'npx.cmd', {}, 'C:\\cwd', true, false, {
  directInvoke: {
    nodePath: 'C:\\Program Files\\nodejs\\node.exe',
    npxScript: 'C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npx-cli.js',
  },
});
```

**问题说明**：`spawnNpxBackend` 的 options 类型现在只有 `{ extraArgs?: string[]; detached?: boolean }`，不再有 `directInvoke`。传入 `directInvoke` 会触发 TypeScript excess property check 错误（tsc 已验证：`error TS2353: Object literal may only specify known properties, and 'directInvoke' does not exist`）。虽然运行时不报错（JavaScript 会忽略多余属性），但会导致严格模式下 tsc check 产生 type error。

**修复建议**：将这两个测试的名称更新为描述新行为，并移除 directInvoke 参数：

```ts
it('uses bundled bun command with chcp prefix on Windows (no special handling)', () => {
  spawnNpxBackend('claude', '@pkg/cli@1.0.0', 'npx.cmd', {}, 'C:\\cwd', true, false);

  const [command] = mockSpawn.mock.calls[0];
  expect(command).toBe('chcp 65001 >/dev/null && npx.cmd');
});

it('uses bundled bun command directly on non-Windows', () => {
  spawnNpxBackend('claude', '@pkg/cli@1.0.0', '/usr/local/bin/npx', {}, '/cwd', false, false);

  const [command] = mockSpawn.mock.calls[0];
  expect(command).toBe('/usr/local/bin/npx');
});
```

---

#### 🔵 LOW — 测试文件中存在不再使用的 stale mock（`getNpxCacheDir`、`safeExec`、`resolveNpxDirect`）

**文件**：`tests/unit/process/services/mcpProtocol.test.ts`，第 28、35–37 行；`tests/unit/acpConnectors.test.ts`，第 41、50 行

**问题代码**：

```ts
// mcpProtocol.test.ts - McpProtocol.ts 已移除这两个导入
getNpxCacheDir: vi.fn().mockReturnValue('/tmp/npx-cache'),

vi.mock('@process/utils/safeExec', () => ({
  safeExec: vi.fn().mockResolvedValue({ stdout: '', stderr: '' }),
}));

// acpConnectors.test.ts - acpConnectors.ts 已不再使用这两个
getNpxCacheDir: vi.fn(() => '/mock-npm-cache/_npx'),
resolveNpxDirect: vi.fn(() => null),
```

**问题说明**：`McpProtocol.ts` 已移除对 `getNpxCacheDir` 和 `safeExec` 的导入，`acpConnectors.ts` 已移除对 `resolveNpxDirect` 和 `getNpxCacheDir` 的使用。这些 mock 是残留代码，不影响测试运行，但会误导读者认为这些函数仍被相关模块使用。

**修复建议**：从对应测试文件的 shellEnv mock 中删除 `getNpxCacheDir`、`resolveNpxDirect` 条目，以及 `mcpProtocol.test.ts` 中整个 `vi.mock('@process/utils/safeExec', ...)` 块。

---

#### 🔵 LOW — test 描述名称未反映新行为

**文件**：`tests/unit/shellEnv.test.ts`，第 419 行

**问题代码**：

```ts
it('prefers the bundled bun binary on Windows', async () => {
  Object.defineProperty(process, 'platform', { value: 'win32' });
  const { resolveNpxPath } = await import('@process/utils/shellEnv');
  const result = resolveNpxPath({ PATH: '/tooling' });
  expect(result.endsWith('bun.exe')).toBe(true);
});
```

**问题说明**：该测试没有 mock `existsSync` 为 true，所以 `getBundledBunDir()` 在测试环境下返回 null，实际上走的是 fallback 路径（返回 bare `'bun.exe'`），而不是 bundled bun 路径。测试名称说"prefers the bundled bun binary"，但实际只覆盖了 fallback，与下方 `falls back to bun.exe when bundled bun is unavailable on Windows` 行为等同。建议增加一个 mock `existsSync` 为 true 的用例验证 bundled bun 路径。

---

### 汇总

| #   | 严重级别    | 文件        | 问题 |
| --- | ----------- | ----------- | ---- |
| 1   | 🟡 MEDIUM | `tests/unit/acpConnectors.test.ts:145,166` | directInvoke 传入已不接受该属性的函数，触发 TS2353 excess property error |
| 2   | 🔵 LOW | `tests/unit/process/services/mcpProtocol.test.ts:28,35` `tests/unit/acpConnectors.test.ts:41,50` | stale mock（getNpxCacheDir、safeExec、resolveNpxDirect）未随功能删除 |
| 3   | 🔵 LOW | `tests/unit/shellEnv.test.ts:419` | 测试名称描述与实际测试行为不符（fallback 而非 bundled bun 偏好） |

### 结论

⚠️ **有条件批准** — 存在小问题，处理后可合并

MEDIUM 问题（directInvoke 导致 TS excess property 错误）需要修复；两个 LOW 问题为清理性改动，建议一并处理。核心逻辑和架构方向正确，bundled bun 简化效果显著。

---

_本报告由本地 `pr-review` skill 生成，包含完整项目上下文，无截断限制。_

--- a/tests/unit/acpConnectors.test.ts
+++ b/tests/unit/acpConnectors.test.ts
@@ -142,13 +142,8 @@ describe('spawnNpxBackend - Windows UTF-8 fix', () => {
     expect(mockChild.unref).not.toHaveBeenCalled();
   });
 
-  it('ignores directInvoke on Windows and still uses bundled bun', () => {
-    spawnNpxBackend('claude', '@pkg/cli@1.0.0', 'npx.cmd', {}, 'C:\\cwd', true, false, {
-      directInvoke: {
-        nodePath: 'C:\\Program Files\\nodejs\\node.exe',
-        npxScript: 'C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npx-cli.js',
-      },
-    });
+  it('uses bundled bun command with chcp prefix on Windows', () => {
+    spawnNpxBackend('claude', '@pkg/cli@1.0.0', 'npx.cmd', {}, 'C:\\cwd', true, false);
 
     const [command] = mockSpawn.mock.calls[0];
     expect(command).toBe('chcp 65001 >nul && npx.cmd');
@@ -161,13 +156,8 @@ describe('spawnNpxBackend - Windows UTF-8 fix', () => {
     expect(command).toBe('chcp 65001 >nul && "C:\\nodejs\\npx.cmd"');
   });
 
-  it('ignores directInvoke on non-Windows', () => {
-    spawnNpxBackend('claude', '@pkg/cli@1.0.0', '/usr/local/bin/npx', {}, '/cwd', false, false, {
-      directInvoke: {
-        nodePath: '/usr/local/bin/node',
-        npxScript: '/usr/local/lib/node_modules/npm/bin/npx-cli.js',
-      },
-    });
+  it('uses bundled bun command directly on non-Windows', () => {
+    spawnNpxBackend('claude', '@pkg/cli@1.0.0', '/usr/local/bin/npx', {}, '/cwd', false, false);
 
     const [command] = mockSpawn.mock.calls[0];
     expect(command).toBe('/usr/local/bin/npx');

